lib/model/query/datasets.js

```diff
@@ -0,0 +1,241 @@
+const { unjoiner } = require('../../util/db');
+const { Dataset, Form } = require('../frames');
+const { map, reduce, compose, pickBy, startsWith, nthArg, assoc, keys, curry, nth, isEmpty, isNil, either, reduceBy } = require('ramda');
+const { construct } = require('../../util/util');
+
+// It removes prefix from all the key of an object
+const removePrefix = curry((prefix, obj) => compose(reduce((acc, key) => assoc(key.replace(prefix, ''), obj[key], acc), {}), keys)(obj));
+
+const pickFrameFields = (frame, obj) => compose(
+  removePrefix(`${frame.def.from}!`),
+  pickBy(compose(startsWith(`${frame.def.from}!`), nthArg(1)))
+)(obj);
+
+const makeHierarchy = reduce((result, item) => {
+  const dataset = new Dataset(pickFrameFields(Dataset, item)).forApi();
+  const property = new Dataset.Property(pickFrameFields(Dataset.Property, item)).forApi();
+  const propertyField = new Dataset.PropertyField(pickFrameFields(Dataset.PropertyField, item)).forApi();
+
+  return {
+    ...result,
+    [dataset.id]: {
+      ...dataset,
+      properties: !property.id ? { ...(result[dataset.id]?.properties || {}) } : {
+        ...(result[dataset.id]?.properties || {}),
+        [property.id]: {
+          ...property,
+          fields: [
+            ...(result[dataset.id]?.properties[property.id]?.fields || []),
+            propertyField
+          ]
+        }
+      }
+    }
+  };
+}, {});
+
+const asArray = compose(map(d => ({ ...d, properties: Object.values(d.properties) })), Object.values);
+
+// should be moved to util.js or we already have similar func somewhere?
+const isNilOrEmpty = either(isNil, isEmpty);
+
+////////////////////////////////////////////////////////////////////////////////
+// SQL Queries
+const _getAllByProjectId = (projectId) => sql`
+  SELECT DISTINCT d.* FROM datasets d 
+  JOIN dataset_form_defs dd ON d.id = dd."datasetId" 
+  JOIN form_defs fd ON fd.id = dd."formDefId" 
+  WHERE fd."publishedAt" IS NOT NULL AND d."projectId" = ${projectId}
+`;
+
+const _insertDatasetDef = (dataset, acteeId, withDsDefsCTE) => sql`
+  WITH ds AS (
+    INSERT INTO datasets (id, name, "projectId", "revisionNumber", "createdAt", "acteeId")
+    VALUES (nextval(pg_get_serial_sequence('datasets', 'id')), ${dataset.name}, ${dataset.projectId}, 0, clock_timestamp(), ${acteeId})
+    ON CONFLICT ON CONSTRAINT datasets_name_projectid_unique 
+    DO UPDATE SET "revisionNumber" = datasets."revisionNumber" + 1
+    RETURNING *
+  )
+  ${raw(withDsDefsCTE ? ', ds_defs AS (' : '')}
+    INSERT INTO dataset_form_defs
+    SELECT ds.id, ${dataset.aux.formDefId} FROM ds
+    ON CONFLICT ON CONSTRAINT dataset_form_defs_datasetid_formdefid_unique 
+    DO NOTHING
+  ${raw(withDsDefsCTE ? '),' : '')}
+`;
+
+const _createOrMerge = (dataset, fields, acteeId) => (isNilOrEmpty(fields) ? _insertDatasetDef(dataset, acteeId, false) : sql`
+${_insertDatasetDef(dataset, acteeId, true)}
+fields("propertyName", "formDefId", path) AS (VALUES      
```

Comment:
It feels a little confusing to me to use the name of an existing table here. How about something like `new_fields` or `new_property_fields`?

Reply:
Not important